### DISCUSSION TITLE
posts/cpp-tricks

### DISCUSSION THREAD

**Discussion by giscus** [EarthMessenger/blog.earthmessenger.xyz#14] [giscus]
# 2022/07/18/cpp-tricks

计算 $\mathrm{highbit}$

https://blog.earthmessenger.xyz/2022/07/18/cpp-tricks.html

**Comment by robinyqc** [comment]
补充一下：g++ 的 priority_queue 实现利用的是 `std::push_heap` 和 `std::pop_heap`。详情可见源码。

关于 priority_queue 能否 O(n) 建堆：可以！`std::priority_queue` 的构造函数采用的是 `std::make_heap` ，因此同样可以 O(n) 建堆。因此 priority_queue 完全可以成为 heap 的安全平替。

**Comment by ForgotDream** [comment]
捉虫：`std::ranges` 里边 `lower_bound` 下边写的是 `unique`
